[#7] Mock 데이터를 이용한 채팅방 UI 로직 구현

kjheee opened the pull request:
## PR 타입
- [X] 기능 개발
- [X] 패키지 추가

## 작업 내용
1. Button 공통 컴포넌트
2. Lines 페이지 구조 설계 및 관련 컴포넌트 생성
3. SidePanel
   - 채팅방 정보
   - 참여 멤버
         - 방장 및 본인 표시
4. ChatPanel
   - 메시지 입력창
         - 메시지 입력에 따라 높이 조정
         - Enter키로 메시지 전송
   - 메시지창
         - `moment`를 이용해 발송 시간 표시
         - 내가 보낸 메세지 스타일 분리
5. Design
   - 브라우저 사이즈에 따라 레이아웃 조정

## 전달사항
- 비동기 데이터 상태 처리는 이후 서버와 실시간 통신을 하면서 tanstack query로 적용할 예정입니다.
- react-hook-form 사용 전으로 Enter키 전송과 함께 개선하겠습니다.

## 디렉토리 구조
`src/pages/Lines` 폴더 안에 관련 패널 폴더를 생성해 구성했습니다.

![issue7_2](https://github.com/f-lab-edu/debus/assets/93242226/244ec1cf-885d-43cd-b543-f92dfbebd41c)


## 실행화면
https://github.com/f-lab-edu/debus/assets/93242226/d1414d2b-ee7a-4b59-9aa0-f693b924633c

## 관련 이슈
close #7 #8

kjheee committed:
#7 design: 채팅방 SidePanel Color 추가

kjheee committed:
#7 refactor: 정류장 라우팅 경로 수정

kjheee committed:
feat: Button 공통 컴포넌트 작성

kjheee committed:
design: divider CSS 작성

kjheee committed:
#7 feat: Lines 컴포넌트 및 스타일 추가

kjheee committed:
#7 feat: SidePanel 컴포넌트 및 스타일 추가

kjheee committed:
#7 feat: ChatPanel 컴포넌트 및 스타일 추가

kjheee committed:
#8 feat: 채팅방 atom 생성 및 더미데이터를 이용한 SidePanel 구현

kjheee committed:
chore: printWidth 수정 및 eslint 규칙 추가

kjheee committed:
#8 feat: ChatForm 기능 구현 및 스타일 추가

kjheee committed:
#8 feat: 더미데이터를 이용한 ChatPanel 구현

- 메세지 불러오기
- 화면 표시

kjheee committed:
chore: moment 설치 및 App 컴포넌트 적용

kjheee committed:
#8 feat: 더미데이터를 이용한 ChatMessages UI 로직 구현

kjheee committed:
#8 design: 브라우저 사이즈에 따라 width/height 설정 변경

kjheee committed:
#8 feat: Enter키로 form submit 기능 추가

kjheee committed:
#8 refactor: 임시 user 데이터 적용 및 timestamp 숫자 변환

kjheee committed:
#8 design: Lines 내부 패널 Height 고정값 수정

- 부자연스러운 움직임 개선

kjheee committed:
#8 refactor: react-hook-form 도입 및 메시지 입력창 적용

kjheee commented:
## 개선사항
Controller를 사용하지 않고 react-hook-form을 적용했습니다.

- react-hook-form
    - `watch`: 입력된 값을 추적해 height과 active 스타일 구현
    - `setFocus`: 채팅방 진입 시 커서 표시
- 빈 문자열 상태에서 submit 적용안함
- Enter submit
    - 기본 줄바꿈 기능 적용안함
    - 분기문 정리

f-lab-jake commented on src/components/UI/Button/Button.tsx:
```diff
@@ -0,0 +1,19 @@
+import styles from './Button.module.css';
+const cx = classnames.bind(styles);
+
+type ButtonPropsType = {
```
뒤에 `Type` 뺍시다. 어차피 타입인건 알잖아요?

f-lab-jake commented on src/components/UI/Button/Button.tsx:
```diff
@@ -0,0 +1,19 @@
+        </button>
+    );
+};
+
```
arrow function을 쓸수있을땐 쓰세요. vs code에서 ctrl + .으로 쉽게 변환할수 있습니다

kjheee replied:
적용되어 있는데 사용법이 잘못되었을까요?!

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatForm/ChatForm.tsx:
```diff
@@ -0,0 +1,83 @@
+        reset();
+    };
+
+    const handleEnterSubmit: KeyboardEventHandler<HTMLTextAreaElement> = (e) => {
```
요런거 `KeyboardEventHandler<HTMLTextAreaElement>` 타입 찾아서 해주는거 참잘했습니다.

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatForm/ChatForm.tsx:
```diff
@@ -0,0 +1,83 @@
+
+const ChatForm = () => {
+    const { register, handleSubmit, setFocus, reset, watch } = useForm<FormData>();
+    const [textareaHeight, setTextareaHeight] = useState('17px');
```
`number`타입을 쓰세요

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatForm/ChatForm.tsx:
```diff
@@ -0,0 +1,83 @@
+    }, [value, resizeTextareaHeight]);
+
+    useEffect(() => {
+        setFocus('content');
```
의도가 뭐지요?

kjheee replied:
페이지에 진입했을때 입력창에 포커스를 두기 위해 사용했습니다. 첫 마운트때만 필요하므로 빈 배열로 수정하였습니다!

f-lab-jake replied:
`autofocus`로 하면 문제가 있나요?

kjheee replied:
가능합니다 😱. autofocus로 해결했습니다.

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatForm/ChatForm.tsx:
```diff
@@ -0,0 +1,83 @@
+    );
+
+    useEffect(() => {
+        if (value) resizeTextareaHeight(value);
```
이거 `onChange`에 넣는게 빠릅니다

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatPanel.tsx:
```diff
@@ -0,0 +1,39 @@
+const ChatPanel = () => {
+    const [messages, setMessages] = useState<Message[]>([]);
+    useEffect(() => {
+        fetch('/data/messages.json')
+            .then((res) => res.json())
+            .then((data) => setMessages(data));
```
cleanup이 필요합니다

kjheee replied:
useEffect 의존성 배열이 빈 배열이면 마운트때만 호출되는데도 필요할까요?

f-lab-jake replied:
fetching 중간에 컴포넌트가 unmount되고, 그다음에 `setMessages`를 하면 warning이 뜹니다.

kjheee committed:
#8 refactor: 메시지 입력창 개선

- textareaHeight state number 타입 적용
- 'undefined' submit 방지
- resizeTextareaHeight onChange에 바로 적용
- setFocus 의존성 배열 수정

kjheee committed:
#8 style: 타입명 수정

kjheee commented:
## 개선사항

1.  타입명에 'Type' 제거
2.  `textareaHeight`: number 타입 적용 후 사용하는 곳에 ‘px' 추가
3.  `resizeTextareaHeight` 함수 onChange에 적용
    - watch로 관찰중인 value는 반박자 느려서 getValues 이용
4. `setFocus`를 작성한 useEffect 빈 배열 에러 처리

kjheee committed:
Merge branch 'main' into feature/#7

kjheee committed:
chore: vite-plugin-svgr 재설치

f-lab-jake commented on src/pages/Lines/ChatPanel/ChatMessages/ChatMessages.tsx:
```diff
@@ -0,0 +1,33 @@
+};
+
+const ChatMessages = ({ message }: PropsType) => {
+    const isMyMessage = useMemo(() => user && user.uid === message.user.id, [message.user.id]);
```
요런건 굳이 `useMemo`안해도 됩니다. 원시 타입이고 계산도 싸니까요.

kjheee committed:
#7 refactor: 원시 타입 useMemo 제거

kjheee committed:
#7 refactor: autoFocus 적용 후 setFocus 제거

kjheee committed:
chore: firebase 재설치

kjheee committed:
#7 refactor: useEffect fetching cleanup

kjheee reviewed (commented):
## 개선사항

- boolean 타입의 `isMounted`를 이용해 마운트 유무로 분기했습니다. 
- 간단히 Loading, Error 처리도 적용했는데 이후 `TanStack-Query`로 리팩토링 하여 정리하겠습니다.
